ulib/FStar.Interactive.PostProcess.fst

```diff
@@ -0,0 +1,760 @@
+let split_assert_conjs dbg res =
+  let ge0 = res.ge in
+  (* Simplify the term (it may be an abstraction applied to some parameters) *)
+  let t = norm_term_env ge0.env simpl_norm_steps res.res in
+  (* Split the conjunctions *)
+  let conjs = split_conjunctions t in
+  (* If there is only one conjunction, check if it is of the following form
+   * and try to split:
+   * [> let Construct x1 .. xn = x in A1 /\ ... /\ Am
+   *)
+  let conjs =
+    if List.length conjs = 1 then split_conjunctions_under_match dbg t
+    else conjs
+  in
+  let asserts = mk_assertions conjs [] in
+  (* Print *)
+  printout_success ge0 asserts
+
+[@plugin]
+val pp_split_assert_conjs : bool -> unit -> Tac unit
+let pp_split_assert_conjs dbg () =
+  try
+    let res = find_focused_assert_in_current_goal dbg in
+    split_assert_conjs dbg res;
+    end_proof ()
+  with | MetaAnalysis msg -> printout_failure msg; end_proof ()
+       | err -> (* Shouldn't happen, so transmit the exception for debugging *) raise err
+
+(*** Term unfolding in assert *)
+/// Unfold/rewrite a term in an assert.
+/// If the term is a (recursive) top-level identifier, unfold it.
+/// Otherwise look for an equality or a pure let-binding to replace it with.
+/// If the assert is an equality, unfold/rewrite only on the side chosen by the user.
+
+// TODO: use "kind" keyword rather than type above
+/// An equality kind
+noeq type eq_kind =
+  | Eq_Dec    : typ -> eq_kind          (* =   *)
+  | Eq_Undec  : typ -> eq_kind          (* ==  *)
+  | Eq_Hetero : typ -> typ -> eq_kind   (* === *)
+
+/// Deconstruct an equality
+// We use our own construct because ``term_as_formula`` doesn't always work
+// TODO: update ``term_as_formula``
+val is_eq : bool -> term -> Tac (option (eq_kind & term & term))
+let is_eq dbg t =
+  let t = remove_b2t t in
+  print_dbg dbg ("[> is_eq: " ^ term_to_string t);
+  let hd, params = collect_app t in
+  print_dbg dbg ("- hd:\n" ^ term_to_string hd);
+  print_dbg dbg ("- parameters:\n" ^ list_to_string (fun (x, y) -> term_to_string x) params);
+  match inspect hd with
+  | Tv_FVar fv ->
+    let name = flatten_name (inspect_fv fv) in
+    begin match params with
+    | [(a,Q_Implicit);(x,Q_Explicit);(y,Q_Explicit)] ->
+      if name = "Prims.op_Equality" || name = "Prims.equals" || name = "Prims.op_Equals" then
+        Some ((Eq_Dec a), x, y)
+      else if name = "Prims.eq2" || name = "Prims.op_Equals_Equals" then
+        Some ((Eq_Undec a), x, y)
+      else None
+    | [(a,Q_Implicit);(b,Q_Implicit);(x,Q_Explicit);(y,Q_Explicit)] ->
+      if name = "Prims.eq3" || name = "Prims.op_Equals_Equals_Equals" then
+        Some ((Eq_Hetero a b), x, y)
+      else None
+    | _ -> None
+    end
+  | _ -> None
+
+/// Reconstruct an equality
+val mk_eq : eq_kind -> term -> term -> Tot term
+let mk_eq k t1 t2 =
+  match k with
+  | Eq_Dec ty ->
+    mk_app (`Prims.op_Equality) [(ty, Q_Implicit); (t1, Q_Explicit); (t2, Q_Explicit)]
```

Comment:
Thanks for the hint! Actually, I used `Prims.op_Equality` on purpose here, because I think equality can be desugared to different identifiers (I think? I saw eq1 and eq2 somewhere for instance) so I wanted to remove ambiguity

Reply:
Hm, `(=)` should be sugar for `op_Equals` (not `op_Equality`), but I couldn't find where that connection is made. So yes, maybe there's something funny going on.